src/Core/Tasks/TaskQueue.cpp

```diff
@@ -178,82 +184,91 @@ void TaskQueue::detectCycles() {
 
 void TaskQueue::startTasks() {
     using namespace Ra::Core::Utils;
-    if ( m_workerThreads.empty() ) {
-        LOG( logERROR ) << "TaskQueue as 0 threads, could not start tasks in parallel. Either "
-                           "create a task queue with more threads, or use runTasksInThisThread";
-        return;
+    {
+        rlock lock( m_mutex );
+        if ( m_workerThreads.empty() ) {
+            LOG( logERROR ) << "TaskQueue as 0 threads, could not start tasks in parallel. Either "
+                               "create a task queue with more threads, or use runTasksInThisThread";
+            return;
+        }
     }
-
     // Add pending dependencies.
     resolveDependencies();
 
     // Do a debug check
     detectCycles();
 
     // Enqueue all tasks with no dependencies.
-    for ( uint t = 0; t < m_tasks.size(); ++t ) {
-        // only queue non null m_tasks
-        if ( m_tasks[t] && m_remainingDependencies[t] == 0 ) { queueTask( TaskId { t } ); }
-    }
+    {
+        wlock lock( m_mutex );
+        for ( uint t = 0; t < m_tasks.size(); ++t ) {
+            // only queue non null m_tasks
 
+            if ( m_tasks[t] && m_remainingDependencies[t] == 0 ) { queueTask( TaskId { t } ); }
+        }
+    }
     // Wake up all threads.
     m_threadNotifier.notify_all();
 }
 
 void TaskQueue::runTasksInThisThread() {
+    // this method should not be called between startTasks/waitForTasks, we do not lock anything
+    // here.
 
-    // lock task queue so no other worker can start working while this thread do the job.
-    std::lock_guard<std::mutex> lock( m_taskQueueMutex );
+    // use local task queue, preventing workers to pickup jobs.
+    std::deque<TaskId> taskQueue;
 
     // Add pending dependencies.
     resolveDependencies();
 
     // Do a debug check
     detectCycles();
-
-    // Enqueue all tasks with no dependencies.
-    for ( uint t = 0; t < m_tasks.size(); ++t ) {
-        // only queue non null m_tasks
-        if ( m_tasks[t] && m_remainingDependencies[t] == 0 ) { queueTask( TaskId { t } ); }
+    {
+        // Enqueue all tasks with no dependencies.
```

Comment:
Don't we need a `rlock` here ?

Reply:
I don't think so. See previous discussion about `TaskQueue::runTasksInThisThread()` on the PR [Refactor textures](https://github.com/STORM-IRIT/Radium-Engine/pull/1056#top) #1056

Reply:
Right, there is the comment atop 
 // this method should not be called between startTasks/waitForTasks, we do not lock anything
    // here.
 
This method also calls  resolveDependencies();  and detectCycles(); which locks, so we can't lock at the beginning of the function.

rlock is an option, but do not mean anything there. A futur version may resolve concurrent/mono thread better, follow up issue #1065